### ISSUE TITLE
Minimize power consumption

### ISSUE THREAD

**Issue by biemster** [biemster/st17h66_FindMy#7]
The current code is based on the SDK v3.1.1.2 lowPower beacon example, which seems to last on a coin cell for a couple/few weeks. There is also a git repo with a power optimized firmware here: https://github.com/ThuanLeUte/ble-st17h66_fw_power_consumption, and there might be other ideas to lower power consumption (like putting the chip in deep sleep and just wake up for a single broadcast every 5 sec).
First it's necessary to actually measure the power consumption of the different firmwares, and then start experimenting with the code to get it lower.

**Comment by biemster** [comment]
@vadimkozhin did you ever analyse the power draw with the rigol on your bench?

**Comment by vadimkozhin** [comment]
> did you ever analyse the power draw with the rigol on your bench?

@biemster Yes. It is aligning with datasheet - 4.7mA during transmit. For now I can't get anything less from a chip, so I decided just plug in the new batt and leave it to check if it lasts more then a week...

So far it works from 16.08. It transmits once in 5 sec.

**Comment by biemster** [comment]
Thanks! Did you also measure the current when idle?

**Comment by biemster** [comment]
I hope I will be able to improve on that, since the tags I got usually had the battery in place, and were switched on by pressing the button. I guess those tags are lying on the shelf sometimes for a month or more, which would drain the battery in that case.

**Comment by vadimkozhin** [comment]
Yes, according to datasheet there is a sleep mode with less current, but it is deactivated only with a hardware pin e.g. with button preset. 
![image](https://user-images.githubusercontent.com/8769702/192284235-a257a659-a02c-4876-856c-abf97337adf1.png)

There is (according to datasheet) couple of interesting options, but I cannot succeed to activate them, with provided SDK samples. Good luck with experiments! :)

**Comment by biemster** [comment]
Thinking about it, 5mA on a 230 mAh 2032 should only give you 40 hours of running, so I guess it is idling already on a lower power mode?

**Comment by vadimkozhin** [comment]
> 5mA on a 230 mAh 2032 should only give you 40 hours of running, so I guess it is idling already on a lower power mode?

Exactly my thoughts! So that's why I decided just plug in the battery and see what will happen after a week or two. So far it is on since 16.08, so yes, definitely idling. May be need to find a way to accurately measure a batt consumption.

**Comment by biemster** [comment]
I had to replace the battery in one of my tags already for the second time, so I'm getting about a month or so, give or take a week. That's shorter than I was hoping..

**Comment by biemster** [comment]
So I put in two fresh batteries when I opened this ticket, but they both seem to have died already 4 days ago. That puts it on just over two weeks with an advertisement every 3 seconds, which is a bit disappointing I must say.
Maybe I should prioritize this.

**Comment by vadimkozhin** [comment]
Mine is two months and still appears on the map. 8 sec interval. Will wait till it drain the battery.

**Comment by biemster** [comment]
Well done! I'll have to think out a way to accurately measure the power draw first. All I have at the moment is a basic multimeter from Lidl, I should be able to do better than that. Also the 2032s I'm using are not of the best quality, so my measurement could have been a bit biased.

**Comment by biemster** [comment]
@vadimkozhin do you by any chance know how long it takes the mcu to send out the advertisement packet?

**Comment by vadimkozhin** [comment]
> do you by any chance know how long it takes the mcu to send out the advertisement packet?

Unfortunately not. The tag is "traveling" at the moment :) If you still will be interested in this interval, ping me in a week or so, then tag will be again in my hands and I can measure actual interval with oscilloscope.

**Comment by olivluca** [comment]
I don't know if it's useful or not, but maybe looking at [this project](https://github.com/pvvx/ATC_MiThermometer) and asking @pvvx could give you some ideas: the battery lasts around 1 year. Looking at the datasheet the telink chip is more efficient (deep sleep 1uA with ram retention vs. 4uA for the st17h66, transmission 4.8mA vs 8.6mA) but OTOH it has to manage a display and the temperature/humidity sensor.

**Comment by biemster** [comment]
I've been eyeing that project for a while as well, especially since the 17h**2**6 seems to be a clone of the telink used in those thermometers. @pvvx seems to be very able with such material, good idea to ask him for suggestions.

**Comment by pvvx** [comment]
I didn't get good results with the available examples from the SDK for PHY62xx and [AiThinker on PB-02-Kit](https://docs.ai-thinker.com/en/blue_tooth_pb) modules (PHY6212).

---

The best example is with [AT-BLE-UART](https://aithinker.blog.csdn.net/article/details/109474628):
Before the test, enter the commands: 
_AT+BLEMAC=234567123456
AT+BLEADVINTV=2560
AT+RST
reboot...
AT+SLEEP=0_

Adv-Interval: 2560*0.625 = 1600 mc.
The result of measuring the current along the 3.3V line only to the module:
The current averaged over 140 seconds is 40.7 μA.
![image](https://user-images.githubusercontent.com/12629515/197289886-c1a6518e-2fc9-43c7-9b35-c31110ef119f.png)
The current in sleep mode is about 19..20 uA, which is a lot.
Usually the sleep current is highly dependent on the software implementation and not on the directions in the datasheet...

**Comment by biemster** [comment]
Thanks @pvvx ! That's very useful info. With 40uA average you'd get 230 days out of a 230mAh button, that's considerably better than the two weeks I'm getting now, and it could even be stretched to a year with a longer adv interval.
I'm still thinking out how to do the power measurement, and get a nice plot like yours, without breaking the bank on new equipment. (although I could just take an interval of 10+ seconds, and measure the sleep current with my DM as well I just realize..)
I assume you made this plot with a scope?

**Comment by pvvx** [comment]
> With 40uA average you'd get 230 days out of a 230mAh button, that's considerably better than the two weeks I'm getting now, and it could even be stretched to a year with a longer adv interval.

These are very bad indicators.
For a key on a typical BLE SoC with advertisement every 3 seconds, the total average consumption should be less than 10 uA.

In active mode for most BLE SoCs (Telink, nRF, ERF):
wake-up (1..2 ms) and advertising transmission (2 ms) - average current 5..9 mA

In sleep mode with RTC and active 32 kilobytes of RAM - 2..3 uA.

(7mA*3ms+0.003mA*2997ms)/3000ms =0.009997 mA = 10uA

> I'm still thinking out how to do the power measurement, and get a nice plot like yours, without breaking the bank on new equipment. (although I could just take an interval of 10+ seconds, and measure the sleep current with my DM as well I just realize..) I assume you made this plot with a scope?

https://github.com/pvvx/UBIA/tree/master/PowerProfiler

For best sampling results (up to 50..100kHz):
GY-169(INA169) + oscilloscope https://aliexpress.con/item/4000107917726.html
or INA199 + oscilloscope



* Replace the shunt resistor with a switchable set with a maximum resistance of 100 ohms.

**Comment by biemster** [comment]
Just ordered a GY-169

**Comment by pvvx** [comment]
> For a key on a typical BLE SoC with advertisement every 3 seconds, the total average consumption should be less than 10 uA.

Module TB-03F, TLRS8253, test_adv_power1: period adv 3 sec, RF TX **0 dB**, ADV_TYPE_NONCONNECTABLE_UNDIRECTED.
Average power **5.8 uA** at 3.3V.

![image](https://user-images.githubusercontent.com/12629515/197640532-fd183c20-2c73-4f2b-b703-d3a0abeb0369.png)


Module TB-03F, TLRS8253, test_adv_power3: period adv 3 sec, RF TX **+3 dB**, ADV_TYPE_NONCONNECTABLE_UNDIRECTED.
Average power **6.1 uA** at 3.3V.

![image](https://user-images.githubusercontent.com/12629515/197640550-c5ef6b27-6762-4366-a139-063dad6bfbd8.png)

**Comment by biemster** [comment]
Those are great numbers @pvvx ! I should definitely target sub 10uA seeing those results.

**Comment by pvvx** [comment]
> With 40uA average you'd get 230 days out of a 230mAh button

The documentation for PHY62x2 does not indicate power consumption in sleep mode with active-retention RAM. And measurements with examples available in the SDK give disappointing readings - 10..20 uA, depending on the retention RAM size.

**Comment by biemster** [comment]
This is more complicated than I thought. I was hoping there would just be a function in the SDK power manager to call, with an enum or so to specify the sleep mode.
What do you mean by "a memory block with a specific leak in the sleep (0.6V)"?

**Comment by pvvx** [comment]
hal_pwrmgr_RAM_retention(RET_SRAM0|RET_SRAM1|RET_SRAM2);	
hal_pwrmgr_RAM_retention(RET_SRAM0);

**Comment by biemster** [comment]
> I didn't get good results with the available examples from the SDK for PHY62xx and [AiThinker on PB-02-Kit](https://docs.ai-thinker.com/en/blue_tooth_pb) modules (PHY6212).

@pvvx There is a newer version of this SDK including the sources of `rf.lib` where some of the sleep code is located available here:
https://github.com/duanmubingshuai/test
Maybe those can help you get better results?

**Comment by pvvx** [comment]
I tested on the [PB-02-Kit](https://docs.ai-thinker.com/en/blue_tooth_pb) by cutting the conductors on the PCB. But on other circuits, leaks are possible ...
A lot of consumption was due to taken examples from the SDK.
Rewriting everything to my version turned out better:
![PHY6212_PB-02_tst-adv-nc](https://user-images.githubusercontent.com/12629515/198692543-37e03ab0-175f-4e19-9708-5ceb7826adcf.png)

This is PHY6212, and ST17H66 is most likely similar to PHY6252. ST17H66 and PHY6252 have a cache for executing code from FLASH. PHY6212 does not.

**Comment by biemster** [comment]
Got my INA219 today! I don't have an STM32 or TLSR lying around so I can't use the PowerProfiler directly, but I have some ESP's, pico's and arduinos to spare so I'm confident I can build something workable. @pvvx 's repo is a great resource for this, so it shouldn't be long!

**Comment by pvvx** [comment]
The input bias current depends on the voltage at the common inputs. It is stable at a stable temperature and is subtracted from the measurements...
Power Profiler is needed not for measuring the average total current, but for debugging with the aim of obtaining the lowest currents using a visual diagram of work. It is possible to measure the SoC sleep current with a simple tester. And to measure the current during activity - no. The average current is the sum of these currents over time.
If you need a turnkey solution, then there is a more or less option for the price and quality of [nRF Power Profiler Kit 2](https://www.nordicsemi.com/Products/Development-hardware/Power-Profiler-Kit-2).

**Comment by biemster** [comment]
thanks for that explanation @pvvx ! Good to know that those input bias currents can be calibrated out, I'm going forward then with this setup. Unfortunately I don't have an oscilloscope on my bench due to budget constraints, so a turnkey solution is even more out of reach :(
I have a DM here capable of uA measurements, so the plan now is to have the chip sleep for 10+ seconds and measure the power draw with that, and then make the diagram with the INA219 and scale the output so it matches the sleep current measured with the DM.

**Comment by vadimkozhin** [comment]
Guys, just to inform you, my setup still working from August 16.

**Comment by biemster** [comment]
> Guys, just to inform you, my setup still working from August 16.

That's amazing! I have quite a few drain on me already, to a point I'm not putting batteries in anymore. I'm still trying to find time to do the power measurement, seeing yours is still running is encouraging!

**Comment by vadimkozhin** [comment]
@biemster I forgot to mention one, may be important thing -- I am completely remove the buzzer from the board. And just for the info, [beacon time interval in my example](https://github.com/vadimkozhin/st17h66-OHS-tag/blob/main/src/example/ble_peripheral/simpleBlePeripheral/source/simpleBLEPeripheral.c#L48) was set to 8000.

**Comment by biemster** [comment]
good point about the buzzer, as it is connected to uart which is active. I'll keep that in mind when I start on this (hopefully during the holidays), I just need to desolder the shunt on my INA219 and calibrate it so it shouldn't be that much work

**Comment by sebi5361** [comment]
Here are the measurements I made on my [Lenze st17h66 beacon](https://www.aliexpress.com/item/1005004744404020.html) flashed with @biemster's latest FindMy firmware (commit [5633d40](https://github.com/biemster/FindMy/commit/5633d408a68e615edf15f59649b32922e3042ba8) on Sep 6, 2022) using the [NORDIC Power Profiler Kit II](https://www.nordicsemi.com/Software-and-tools/Development-Tools/Power-Profiler-Kit-2):
![image](https://user-images.githubusercontent.com/18078043/210271720-8d3b649e-1060-4d1b-8114-e9c234ca311e.png)
The average consumption is ~234.47uA @ 3V (spikes are every ~3.127s).
It is 30 times more than to the nRF51822 FindMy beacons [I have tested](https://github.com/seemoo-lab/openhaystack/issues/57#issuecomment-880210121) (234uA vs 7.60uA), so I think there is a great margin for improvement... Tell me if I can help with additional measurements.

**Comment by biemster** [comment]
that's great work @sebi5361 ! I noticed my batteries draining quite fast as well, this confirms that. I'm on holiday atm so i can't work on this for a couple days, but I'm definitely going to ask you to run this test again a couple of times on some new versions of the code, if you are willing. thanks!

**Comment by sebi5361** [comment]
Happy to help!

**Comment by biemster** [comment]
@sebi5361 from your measurements, what is the power draw while the chip is sleeping? and how many milliseconds do those peaks last? (I'm trying to figure out where most energy is spent, should i optimize the sleep or the peaks)

**Comment by sebi5361** [comment]
@biemster: I will give you the exact values pretty soon but you should start by optimizing the sleep...
Indeed the average consumption is close to the sleep consumption (horizontal blue line on the graph is close to 234uA) currently.

**Comment by sebi5361** [comment]
# Idle
224uA * 3.12s = 700uC
![2023-01-04 04 59 35](https://user-images.githubusercontent.com/18078043/210482478-5084896b-c5a1-430b-bf7e-641a0f7f26dd.png)
# Emission
6.4mA * 4.6ms = 30uC
Note the 3 similar emission spikes.
![2023-01-04 05 01 57](https://user-images.githubusercontent.com/18078043/210482522-dec49786-6b28-4826-a444-2ca7d44e17e5.png)
# Conclusion
700uC vs 30uC = 23 times more reasons to spend time on idle rather than emission at the moment.

**Comment by pvvx** [comment]
3 peaks is a typical transmission of BLE advertising over 3 channels.
Problems only in the first 2 ms. But this is an invariable part - this is the wake-up time of this SoC. A small optimization is possible, by a few percent, but this is already related to the internal algorithms in the SDK implementation.

Sleep consumption is atypical for this SoC. Something is set incorrectly. Most likely GPIO pull-ups or internal controllers are not disabled.

**Comment by vadimkozhin** [comment]
> For what reason might the buzzer be leaking some current? Is it the same for the LED? Should I remove both to perform additional tests?

My version of the tag ticking the buzzer when chip output something on UART. So, if you completely silent the UART or hear no any ticking during packet transmit -- do not worry about the buzzer. But you also can measure if there is some voltage on buzzer during the packet send... 

With LED -- it's the same. If during transmission led blinking, you can remove it (or cut the trace to it) and redo measurements, coz LED definitely eating some power.

**Comment by sebi5361** [comment]
Some great news:
Without the buzzer (+ its related diode  + its related CMOS), nor the LED, the average consumption drops from 234.47uA to 15.6uA @ 3V!
![2023-01-05 18 58 30](https://user-images.githubusercontent.com/18078043/210855463-aee292f3-8ebf-4192-af9d-337213ccf040.png)
<img src="https://user-images.githubusercontent.com/18078043/210855600-9ab29466-6e2e-4286-b408-1ea043aa43b5.JPG" height="400" />
The improvement during the emission phase is modest (27.62uC vs 29.49uC) but becomes significative during idle (idle average current is 6.37uA vs 224.16uA):
![2023-01-05 19 06 30](https://user-images.githubusercontent.com/18078043/210856307-3496784e-293c-45fb-b1d9-2d1caebee7d2.png)
So there is now only a factor 2 for this cheap Lenze st17h66 beacon to become as efficient as a NORDIC nRF8166 beacon (15.6uA vs 7.6uA).
As a nRF8166 beacon lasts ~1 year on a single CR2032 battery (those are real life measurements), one can expect our Lenze st17h66 beacons without buzzers to last ~6 months using the current firmware.
Note: There is already room for improvement by changing the current advertising period from 3.127s to 5s (as with the nRF8166 beacons).

**Comment by biemster** [comment]
wow great find @sebi5361! this is really pinpointing where the code should be fixed. when I'm back from holiday after the weekend I'll get on this definitely. disabling the uart altogether or fixing pullups on the gpio as @pvvx mentioned should hopefully be as effective as removing the buzzer

**Comment by sebi5361** [comment]
Unfortunately I won't be able to test this aspect of the code modification on my Lenze beacon as there is no buzzer anymore...
But if you decide to work on (i) the advertisement period, (ii) idling in sleep mode, (iii) preferably without SRAM retention, I'll be able to perform additional tests easily.

**Comment by pvvx** [comment]
> So there is now only a factor 2 for this cheap Lenze st17h66 beacon to become as efficient as a NORDIC nRF8166 beacon (15.6uA vs 7.6uA).

That's an awful lot. Should be up to 4 uA. Something else is not disabled...
Datasheet: _4uA @ Sleep Mode with 32 KHz RTC and all SRAM retention_

**Comment by sebi5361** [comment]
Hi @biemster, looking forward to knowing if you made any progress?

**Comment by biemster** [comment]
> Hi @biemster, looking forward to knowing if you made any progress?

Not yet, work piled a bit over the holidays. But this is the first thing I pick up when time allows!

**Comment by biemster** [comment]
Just made some initial steps, my DMM shows it consumes 270 uA even! @pvvx I noticed most gpio are initialized floating:
```C
ioinit_cfg_t ioInit[]=
	{
		{GPIO_P02,   GPIO_FLOATING   },
		{GPIO_P03,   GPIO_FLOATING   },
		{GPIO_P07,   GPIO_FLOATING   },
		{GPIO_P09,   GPIO_PULL_UP    },
		{GPIO_P10,   GPIO_PULL_UP    },
		{GPIO_P11,   GPIO_FLOATING   },
		{GPIO_P14,   GPIO_FLOATING   },
		{GPIO_P15,   GPIO_FLOATING   },        
		{GPIO_P18,   GPIO_FLOATING   },
		{GPIO_P20,   GPIO_FLOATING   },
		{GPIO_P34,   GPIO_FLOATING   },
```
except for the UART, is that the correct way? (next I'll disable the uart pins here too, to see if that makes a difference)

**Comment by pvvx** [comment]
I have not yet received the described devices. Stuck somewhere in the mail...

**Comment by biemster** [comment]
Wow changing the gpio init on pins 9 and 10 from pullup to floating reduced the power draw from 270uA to only 6uA! That's a 45x reduction!
Thanks @pvvx and @sebi5361 , your inputs were vital on this one. I'm going to experiment a bit more on the SRAM retention, but we tackled the big one.

**Comment by biemster** [comment]
Playing with the SRAM retention was not successful yet. If I disable it completely, or retain only SRAM1 or 2, the power draw is initially below 2 uA but after the first advertisement it seems to shoot through the roof..

**Comment by biemster** [comment]
Enabling the low current LDO works though! The SDK seems to indicate that only the 17h6**5** supports this, but activating this on my 17h66 reduces the draw from 6uA to 3.3uA! @sebi5361 could you run a test with the new version from https://github.com/biemster/FindMy/tree/main/Lenze_ST17H66 please? (note that the flasher script changed, since the key is not in the same location anymore) Thanks!

**Comment by sebi5361** [comment]
@biemster: here are the results with the new firmware:
- The average consumption is now 11.97uA (previously it was 15.6uA)
![2023-01-17 17 32 31](https://user-images.githubusercontent.com/18078043/212957727-d8ddb3b0-fcc4-4cd7-9dc0-65a34f1e1f4e.jpg)
- The consumption during idle is now 3.13uA (previously 6.37uA):
![2023-01-17 17 34 44](https://user-images.githubusercontent.com/18078043/212957738-d67df02d-9bb7-4a04-a17d-57da37e6e239.jpg)

Already some great improvement!
BTW: I ordered a new Lenze st17h66 beacon. As soon as it arrives I will be able to do some tests with the buzzer in place.

**Comment by biemster** [comment]
Great @sebi5361! Thanks for testing this. So with an idle consumption of 3uA and an average of 12uA, the bulk of the power is drawn while transmitting. I don't have any ideas on how to improve this, except the transmit interval. I think I will add an option to the flasher to set this while flashing.

**Comment by sebi5361** [comment]
> I think I will add an option to the flasher to set this while flashing.

Indeed that is a good idea! The broadcast interval is currently 3.127s. Is it as programmed?

Also, I can do some additional tests to help you figure out what is going wrong with the SRAM retention.

**Comment by biemster** [comment]
The broadcast interval is in steps of 625 milliseconds for some reason, 5000 of them in this implementation (=3125, so quite spot on!). On the SRAM retention, there is probably some level of retention necessary otherwise the chip loses all the initializations it did on boot. I guess that's what is going wrong when I completely disable it?

**Comment by pvvx** [comment]
>So there is now only a factor 2 for this cheap Lenze st17h66 beacon to become as efficient as a NORDIC nRF8166 beacon (15.6uA vs 7.6uA).
As a nRF8166 beacon lasts ~1 year on a single CR2032 battery (those are real life measurements), one can expect our Lenze st17h66 beacons without buzzers to last ~6 months using the current firmware.
Note: There is already room for improvement by changing the current advertising period from 3.127s to 5s (as with the nRF8166 beacons).

(4.32x6.93+0.00313x3120)/3127= 0.012696897985... mA = 12.7 uA

4.32mA x 6.93ms-> https://user-images.githubusercontent.com/18078043/210482522-dec49786-6b28-4826-a444-2ca7d44e17e5.png
0.00386mA x 3120ms-> https://user-images.githubusercontent.com/18078043/212957738-d67df02d-9bb7-4a04-a17d-57da37e6e239.jpg

(4.32*6.93+0.00313x**5000**)/5007=0.0091 mA = 9.1 uA

**Comment by pvvx** [comment]
I know it. And still it is suitable, since the average for the period does not change and did not recalculate.
Plus ads have an additional random duration addition and not every active period has the same average due to different code execution. But in general it gives average readings.

**Comment by sebi5361** [comment]
> The graphs show that the device does not have the recommended capacity in the power circuit and CR2032 will not be used up by 100% due to an increase in internal resistance and a voltage drop under a load of 12 mA (!).
Adding a ceramic capacitance of tens of microfarads will give an additional 40% of the operating time from the CR2032.

I'll try some tens of microfarad capacitors to make the broadcasting spikes drop below 12mA, and post the new graphs.

**Comment by pvvx** [comment]
The trouble is not in the current peaks. When loaded, the voltage drops. A run down battery increases internal resistance. The voltage without load remains about 3V. The capacitor charges during sleep and gives back in the active phase, without a large voltage drop.

**Comment by sebi5361** [comment]
Your explanations are really clear. Thanks you!

- What exact value for the capacitor would you recommend?
- I only have ceramic through hole capacitors (not SMT). Would they work? (What about their "equivalent series resistances"?)

I can do a test with the NORDIC Power Profiler II setting the voltage to 1.8V and checking if the added capacitor makes the current remain below 1.2mA. Additionally, this test would allow us to conclude about the capacitor "equivalent series resistance" when looking for an increase in the current at idle, no?

**Comment by biemster** [comment]
> 3 peaks is a typical transmission of BLE advertising over 3 channels.

I'm wondering if the detection probability would also drop 3-fold if the chip advertises on only 1 channel? Anything less than 3x I would consider a gain (although consumption would not decrease that much due to wake-up losses).

**Comment by sebi5361** [comment]
I added a 100ohm resistor in series between the 3V-output NPPK2 (NORDIC Power Profiler Kit II) and the beacon to simulate a CR2032 battery at its end life.
![2023-01-18 19 26 15](https://user-images.githubusercontent.com/18078043/213276097-25c7ca85-31b9-4102-b5ed-50aed8322e95.png)
As can be seen on the above graph, with such an electronic assembly the current peak occurs during wake-up (contrarily to [previous recordings](https://github.com/biemster/st17h66_FindMy/issues/7#issuecomment-1370470843), which I can not explain. However the peak value remains the same). 14mA corresponds to a 1.4V voltage drop along the resistor (100ohm * 0.014A), thus leaving 1.6V only for the beacon for a short period of time (3V - 1.4V), which is lower than the [official 1.8V minimum operating voltage](https://www.lenzetech.com/public/store/pdf/jsggs/ST17H66B2_BLE_SoC_Datasheet_v1.1.2.pdf).

Then, with the same setup, I decreased the NPPK2 voltage output to simulate for cold (poor battery chemical reaction).
![2023-01-18 20 22 27](https://user-images.githubusercontent.com/18078043/213279609-fda08550-9c15-41b0-b4af-6fbd2b0e4da4.png)
As can be seen on the above graph, 2.6V seems to be the lower battery voltage for which the beacon still works, which corresponds to a momentary 1.2V for the beacon (2.6V - 1.4V; 1.4V = 100ohm * 0.014A).

Then, I tried to play with capacitors (from 15pF up to 10uF, in parallel, after the resistor on the beacon side) but didn't manage to decrease the peak magnitudes. I cannot figure out why? I was expecting to find a sweat spot for the capacitor value for which the current peaks would decrease...

**Comment by biemster** [comment]
> Then, I tried to play with capacitors (from 15pF up to 10uF, in parallel, after the resistor on the beacon side) but didn't manage to decrease the peak magnitudes.

How exactly did you add the cap? @pvvx will know better than me, but I would guess you'd have to add it to act like a low pass filter:

![image](https://user-images.githubusercontent.com/5699190/213290046-8248a7c8-ab02-406f-9e0c-0b20ac0896c0.png)

**Comment by sebi5361** [comment]
> How exactly did you add the cap? @pvvx will know better than me, but I would guess you'd have to add it to act like a low pass filter:

Exactly as on your scheme! I will try again later. Maybe I did a mistake...

> Anything less than 3x I would consider a gain (although consumption would not decrease that much due to wake-up losses).

According to my [previous broadcast recording](https://user-images.githubusercontent.com/18078043/213276097-25c7ca85-31b9-4102-b5ed-50aed8322e95.png), the consumption would decrease almost in half (surface below the curve would decrease almost in half) if suppressing the 2 last emissions. But maybe the BLE standard doesn't allow this...

> I don't have any ideas on how to improve this, except the transmit interval. I think I will add an option to the flasher to set this while flashing.

If too much work, maybe just set a 5s broadcasting period, as it is the default value on OpenHaystack...

**Comment by pvvx** [comment]
> As can be seen on the above graph, with such an electronic assembly the current peak occurs during wake-up

The peak is possible due to the shunt switcher in NPPK2.

**Comment by sebi5361** [comment]
> The peak is possible due to the shunt switcher in NPPK2. That's why I mentioned that NPPK2 is not designed for accurate measurements, but is used for debugging low-power applications.

Indeed. A good option to check for this would be to use an oscilloscope to mesure the voltage across the 100ohm resistor. Unfortunately I don't have an oscilloscope on hand...

**Comment by pvvx** [comment]
> Unfortunately I don't have an oscilloscope on hand...

This is no required. And so the average consumption of 9 uA is enough for more than a year of battery life.

**Comment by pvvx** [comment]
For comparison CHM582.

[Broadcaster](https://github.com/openwch/ch583) - BLE advertisement only, no connectivity and response to scans. Those. does not turn on the packet receiver after transmission. Type "Beacon" with an interval of 3 seconds - this is the limit for the successful receipt of a beacon with the lowest consumption and safety from the chip.

Changed in SDK usage: HAL_SLEEP=1, DEBUG - disabled, #define DEFAULT_ADVERTISING_INTERVAL 4800 (3 sec), DC-DC enabled, LED on PB4 disabled.

Final measurements of a copy of the CH582M chip on the YD-CH58x board:

DC-DC enabled.
Sleep mode - 2.7uA (3 sec)
Activity - just under 3ms
Wake-up pulse to turn on the active mode more than 6 mA (with a capacitive decay of ~0.4 ms)
Consumption after switching on - 1.2 mA (~ 1 ms)
Transmit - three bursts at 6..7.5 mA (~1.1 ms, 0dB)

Average consumption from 3.3V with an advertising interval of 3 seconds - 6.75 μA

![image](https://user-images.githubusercontent.com/12629515/213306253-ebececb3-520a-4d8d-9413-ee5c40616a4f.png)

And if the DC-DC is incorrectly initialized in the project, then ~ 30% more consumption is obtained.

**Comment by biemster** [comment]
I'll disable the LOG console, and see if I can find the DC-DC initialization

**Comment by biemster** [comment]
Disabling the LOG console on line 197 did not change the resulting binary, so I guess it was not active and optimized out already by the compiler?

**Comment by biemster** [comment]
So advertising on a single channel is not efficient, both my ESP32 and android miss the advertisement very often. The wakeup draw is going to count extra if the single channel advertisement is missed often.

**Comment by sebi5361** [comment]
To better evaluate the current passing through the beacon, I used an oscilloscope to measure the voltage across a 100ohm shunt resistor, powered in series by a 3V stabilized power supply  (see below).
![IMG_6107](https://user-images.githubusercontent.com/18078043/213546429-5ce9908c-8ad2-425b-baa2-8b658fcc8bc8.JPG)
Note that, as explained by @pvvx, there are no more peaks due to the shunt switcher of the NPPK2.

The maximum voltage across the resistor is about 0.9V, which leaves 2.1V for the beacon (3V - 0.9V) and corresponds to a 9mA current across the circuit (0.9V / 100ohm). It was 14mA previously when estimated using the NPPK2, [due to peaks coming from the shunt switcher](https://github.com/biemster/st17h66_FindMy/issues/7#issuecomment-1387722382).

In the current state, without an additional capacitor, the chemical reaction inside a CR2032 battery can fall down to 2.7V due to cold (+ the internal resistor can reach up to 100ohm due to battery wear) without the beacon voltage dropping below its 1.8V brown-out voltage (2.7V - 0.9V = 1.8V).

**Comment by biemster** [comment]
So this means no extra capacitor is needed for the chip to empty the CR2032 properly?

**Comment by pvvx** [comment]
It is necessary if you want a long device operation from the CR battery. See recommendations from Nordic (nRF) and other manufacturers. The CR2032 has 210mAh at a typical discharge current of 0.2-0.6mA. At high currents, the battery capacity decreases.
[High pulse drain impact on CR2032 coin cell battery capacity](https://www.dmcinfo.com/Portals/0/Blog%20Files/High%20pulse%20drain%20impact%20on%20CR2032%20coin%20cell%20battery%20capacity.pdf)

https://www.ti.com/lit/wp/swra349/swra349.pdf

https://www.avnet.com/wps/portal/abacus/resources/article/extending-battery-lifetime-in-wireless-applications-using-supercapacitors/

...

**Comment by sebi5361** [comment]
I recorded the current consumption of the Lenze beacon running the [latest firmware](https://github.com/biemster/st17h66_FindMy/commit/68f7b2be9ca8ec68f289e885bdb3673ee2179153) connected directly to the NPPK2 (no 100ohm resistor, no capacitor, 3V output) in order to compare the average value with the one of the [NORDIC nRF51822 beacons](https://github.com/seemoo-lab/openhaystack/issues/57#issuecomment-880210121) I tested in the same conditions in the past.
![1 min average](https://user-images.githubusercontent.com/18078043/213782683-123fbef5-5b3e-42f7-bdcb-b85021196984.png)
The average current consumption is 8.85uA which is very close to the 7.60uA of the nRF51822 beacons, thus letting me expect a quite similar life time of about 1 year.

![idle](https://user-images.githubusercontent.com/18078043/213783970-db2bc82a-4b69-40b6-8467-22f60f615c50.png)
The sleep current consumption is 3.21uA.

![broadcating2](https://user-images.githubusercontent.com/18078043/213784628-9b15fc43-5df8-41c4-9e82-0ecb31f68b85.png)
The broadcast sequence occurring on 3 channels, lasts less than 5ms, draws a maximum current of about 10mA (omitting the peaks) and represents an average current of 5.98mA.
I drew with vertical arrows what I believe are peak artefacts representative of shunt resistor switching. Rising arrows (decreasing respectively) correspond to switching to a higher shunt value (lower respectively).

**Comment by biemster** [comment]
Very nice @sebi5361. I actually found a cap on my tag connected like the TI whitepaper @pvvx mentioned, so it seems that on at least some tags they are already included. I agree with your lifetime in the order of a year, so I'm considering closing this ticket as fixed. (We can always open it again if we think of something to try out)

**Comment by sebi5361** [comment]
Yes you might consider closing the ticket. However I don't really understand why SRAM is mandatory for broadcasting the FindMy signal every 5s. Aren't SoC registers and ROM enough for this basic task? (I am still wishing to find a way to avoid SRAM retention, thus decreasing power consumption).

Regarding the capacitor, inspired from the formula from Annexe A from the TI whitepaper, I computed its capacity:
Let us imagine the CR2032 battery in the cold delivering 2.1V à -25C, with an internal resistor of 100ohm as it is its end life.
So the battery in that state can deliver a maximum of 3mA to the SoC without reaching the 1.8V brown out voltage (2.1V - 1.8v  = 0.3V, 0.003A = 0.3V / 100ohm).
On the [graph from the  oscilloscope](https://github.com/biemster/st17h66_FindMy/issues/7#issuecomment-1397547285) I observe that the current exceeds those 3mA during broadcasting (voltage in the 100ohm resistor shunt exceed 0.3V). Thus, to avoid brown out all the current on top of that 3mA horizontal line needs to be provided by the capacitor. I try to estimate the corresponding charge that needs to be provided by the capacitor by estimating the surface area under the curve during broadcasting down to that horizontal 3mA horizontal line (as the surface below can be provided by the battery):
dq = T * dI = 0.005s * 0.003A = 15e-6C, where 5ms correspond to the duration of the broadcasting curve above the battery 3mA mark, and dI = 3mA is an estimate of the average current above that mark during that duration (computed by eye trying to fill voids with peaks).
As q=C*U for a capacitor where C is its capacity in Faraday, and as the capacitor has to deliver the charge dq while its voltage drops from 2.1V to 1.8V (dU = 0.3V), its capacity is obtained:
C = dq / dU = 15e-6C / 0.3V = 50e-6F = 50uF.
Thus, those [6V 47uF capacitor](https://www.aliexpress.com/item/1066760503.html) might do the job helping the beacon to last longer in cold places. Hopefully their leakage current will be small enough to provide a benefit.

**Comment by pvvx** [comment]
For stability of operation in different temperature conditions, an addition is necessary - when measuring the supply voltage below the response threshold, the SoC BOR should switch to deep sleep mode with a period of several minutes.
After this time, the battery voltage may increase and operation will be restored. If the SoC turns off by BOR, it will immediately land the battery with multiple cold starts.

> Yes you might consider closing the ticket. However I don't really understand why SRAM is mandatory for broadcasting the FindMy signal every 5s. 

This refers to the applied algorithms in the SDK.
The time and consumption of the chip during a "cold start" in the SDK is many times greater than waking up from sleep mode.

This issue for "beacon" is solved without using the BLE stack. A completely different approach is needed - initialize the RF part and assign buffers for DMA RF, transmit to 3 channels and switch to deep-sleep with RTC. In this case, the BLE stack and the selected SDK are not needed. Examples of working with the RF part are given in PHY62XX_SDK_3.1.1\example\PhyPlusPhy\smart_rf\source\phy_plus_phy.c

You can measure how the chip starts up with the current SDK on power up. This will be similar to using wake up without saving the SRAM.

**Comment by olivluca** [comment]
Besides, I think SRAM is needed to implement issue #2 (to keep track of the current key and when to switch to the next one).

**Comment by pvvx** [comment]
Once an hour, you can write the key to Flash.
This will not affect the overall consumption of the device in any way.
Writing and erasing a Flash sector of 4 kilobytes once an hour gives an increase in consumption of only 17 nanoAmps.

**Comment by olivluca** [comment]
Actually it would just be the key index (a single byte if there are less than 255 keys), but wouldn't it eventually damage the flash (writing to it constantly)?

**Comment by pvvx** [comment]
Adding a 4-byte value to one Flash sector is 1024 times.
The applied SPI Flash has a guaranteed standard of 100,000 rewrites.
The chip has dozens of never used Flash sectors.
In total, when adding a 4-byte value in Flash once an hour to only one sector:
1024 x 100000 / (24 x 365) = 11689 years before the possible deterioration of data storage in this Flash sector.
That is, after these centuries, data written to this Flash sector will have cell leakage and may deteriorate during less than 25 years guaranteed by the manufacturer.
![image](https://user-images.githubusercontent.com/12629515/214089746-4e9e3578-6791-4d38-81d9-361a0588ff44.png)

**Comment by biemster** [comment]
@pvvx you are like a walking datasheet in these things, great to have you here! Are you planning on an attempt to implement your ```A completely different approach is needed - initialize the RF part and assign buffers for DMA RF, transmit to 3 channels and switch to deep-sleep with RTC.``` idea when your finally receive your devices in the mail?
I'm happy to send you a couple of mine as well, that might arrive faster.

**Comment by pvvx** [comment]
The problems are not in writing the program, but in tests and user support. This requires a lot of time, effort and other means.

**Comment by olivluca** [comment]
@pvvx thank you. So writing the key index to flash doesn't appear to be an issue, but if SRAM retention is already necessary to keep consumption low, it may well be used also to keep the index (I suppose it would be much easier to just do an "keyindex=keyindex+1" than to program a routine to write the key index to flash).

**Comment by biemster** [comment]
> The problems are not in writing the program, but in tests and user support. This requires a lot of time, effort and other means.

@pvvx If you don't mind writing the initial code, I can do the tests and support from there?

> but if SRAM retention is already necessary to keep consumption low

@olivluca it is my understanding that the SRAM retention is actually keeping consumption high(ish)?

**Comment by olivluca** [comment]
@pvvx this is a crazy idea, so feel free not to reply, but would it be possible for the MiThermometer do do double duty? (i.e thermometer and findmy beacon?) A thief would never think of an unassuming termomether, with a smiling or sad face, to be also a tracking device :wink: 

@biemster maybe I misunderstood this sentence: 

> The time and consumption of the chip during a "cold start" in the SDK is many times greater than waking up from sleep mode.

**Comment by biemster** [comment]
> @biemster maybe I misunderstood this sentence:
> 
> > The time and consumption of the chip during a "cold start" in the SDK is many times greater than waking up from sleep mode.

I believe Victor suggested to omit the whole SDK cold start, and just init the RF part. No BLE stack or SDK needed, only RTC to kick the RF to send a raw advertisement.
EDIT: I actually overlooked the code example @pvvx sent with this suggestion, maybe I can have a crack at it myself too!

**Comment by pvvx** [comment]
With "retention RAM" there is another dependency. Part of the fast execution code resides in RAM. Also in RAM are programs for writing Flash. When sleeping with RAM power, this code will not be rewritten from Flash each time. If deep-sleep is used, then it takes time to initialize, rewrite critical code from Flash to RAM, which increases battery consumption. It is required to calculate which of these options is more profitable. Without real practical tests, this cannot be done.

**Comment by sebi5361** [comment]
I'll be happy to help too and measure the power consumption of different firmware candidates!
I gave away my beacon this week-end but I am expecting to receive a new one soon to do the tests.

I understand two different beacon types are considered: (i) a basic one with only a single public key being broadcasted, and (ii) a more advanced one involving several public keys. Power consumption might be reduced, at least for beacon type (i), using the new discussed approach.

**Comment by Itheras** [comment]
"just got a st17h66 I can play with I also have a nordic power profiler. I can gladly help as well."

Actually what I just got is the st17h26. I will try to get the 66 variant

**Comment by biemster** [comment]
> "just got a st17h66 I can play with I also have a nordic power profiler. I can gladly help as well."
> 
> Actually what I just got is the st17h26. I will try to get the 66 variant

If you're shopping on AliExpress, try to see if the description states without battery. I've got a few 26's as well, and they all came without CR2032. All the 66's included a battery. Not foolproof, but it will increase your changes probably.

**Comment by sebi5361** [comment]
> > Actually what I just got is the st17h26. I will try to get the 66 variant
> 
> If you're shopping on AliExpress, try to see if the description states without battery. I've got a few 26's as well, and they all came without CR2032. All the 66's included a battery. Not foolproof, but it will increase your changes probably.

The round one I bought from this [link](https://www.aliexpress.com/item/1005004744404020.html) on AliExpress is a st17h66 and included a battery.

**Comment by biemster** [comment]
Following @pvvx 's idea, I've started a project using the RF only, no SDK or BLE stack:
https://github.com/biemster/st17h66_RF
It's basically the PHY62XX 3.1.1 `smart_rf` example, I did not test anything on it yet. But it compiles, and looks like a lot less code than the current implementation!

**Comment by sebi5361** [comment]
> But it compiles, and looks like a lot less code than the current implementation!

That is great news already!

**Comment by biemster** [comment]
> > But it compiles, and looks like a lot less code than the current implementation!
> 
> That is great news already!

After testing it is also broadcasting the MAC I set myself, this code is looking promising!

**Comment by biemster** [comment]
@sebi5361 if you could do a quick power test with https://github.com/biemster/st17h66_RF/blob/161c1e71ed7da982b10e596b6686b0fb1fa8aa6f/bin/Lenze_rf.hex (although the python flasher will not work on this, so you have to use LeKit), that would be swell. I'm getting odd readings on my DMM after the first advertisement, but the chips seems to be functioning as intended.
EDIT: I figured out that the odd readings come from the RX events, the code listens for advertisements every 100ms. Without the RX events, power consumption between TX events is a stable 3.4uA.

**Comment by sebi5361** [comment]
Can somebody else help for this? (Explanations for using LeKit can be found [here](https://github.com/vadimkozhin/st17h66-OHS-tag/tree/main/flash_tool). It is quite straight forward.)
Indeed I offered my Lentze beacon last week-end to a friend (with the 8.85uA average current firmware) and my new beacon is still in the mail (it didn't even leave China yet...). As soon as I receive it I'll let you know.

**Comment by biemster** [comment]
Ah yeah sorry @sebi5361 , forgot about that. Let's postpone this a bit then, since @Itheras is also still waiting for his 17h66. On the plus side, I've finished the new flasher, just need to test it before I push it.

**Comment by biemster** [comment]
I've added a python flasher:
https://github.com/biemster/st17h66_RF/blob/main/flash_st17h66.py
and temporary disabled RX events so we are able to do a proper comparison with the current implementation in this repo. The hex we should profile is here:
https://github.com/biemster/st17h66_RF/blob/f6d533f434d9bb93f7a3c17b88a12616edae7d99/bin/Lenze_rf.hex

**Comment by biemster** [comment]
Another update, I've abolished the rf.lib binary blob. This reduced the resulting firmware by half, so that might make the wake-up more efficient. Again, the hex we should profile is now here:
https://github.com/biemster/st17h66_RF/blob/38bbb31c94b0d6924ee90918680be5d322abf394/bin/Lenze_rf.hex

**Comment by Itheras** [comment]
I was able to get h66 version. using RF hex.

Average Power Usage 7.76uA
![average](https://user-images.githubusercontent.com/70117539/214970486-2a48bce2-3f30-4bc1-a877-ba521e86d91a.png)

broadcast 5.71mA average
![broadcast](https://user-images.githubusercontent.com/70117539/214970491-5079b05f-38a3-485c-873f-4959c1f25c0f.png)

Sleep 3.12uA average
![sleep](https://user-images.githubusercontent.com/70117539/214970494-05149c56-7d61-46ea-8fa3-614efeb2cc70.png)

It seems there is some slight improvement. Good job biemster and everyone.

**Comment by biemster** [comment]
Awesome, thanks @Itheras ! That looks great. The high thin spikes during broadcast are from the power profiler switching shunts right?

**Comment by Itheras** [comment]
> Awesome, thanks @Itheras ! That looks great. The high thin spikes during broadcast are from the power profiler switching shunts right?

Well I think you are correct.

New cr2477 batt sleep 3v
![batt sleep h66](https://user-images.githubusercontent.com/70117539/215120670-6410827d-5670-499c-b88a-37fede2a5522.png)

battery adv 3v
![h66 batt](https://user-images.githubusercontent.com/70117539/215120678-7b225e0b-7e50-409e-9342-789abd5bea1f.png)

using ppk2 power source on nrf51
![nrf51adv](https://user-images.githubusercontent.com/70117539/215120683-a0126613-4131-46ca-bb9b-3b8de43b2ff4.png)

**Comment by pvvx** [comment]
The PPK2 switching thresholds can be easily checked by connecting a 470..1000 uF capacitor in series with a 33..100 Ohm resistor. And watch the falling charge current.
![image](https://user-images.githubusercontent.com/12629515/215189154-d1c17a80-42af-438f-b80c-2afec8f4282f.png)

**Comment by biemster** [comment]
On the DC-DC, I came across the #define `DCDC_CONFIG_SETTING`, which is set almost everywhere to `0x0a`, except in `rf_phy_driver.c` I find `DCDC_CONFIG_SETTING(0x08); //set dcdc to highest`. @pvvx Do you have any clue how this works? What could the meaning of the parameter be?

**Comment by pvvx** [comment]
Most likely, the ST17H66 chip does not have a DC-DC output and only LDO works. From this, the energy consumption is greater than that of similar chips with DC-DC.
You need to try to disable DC-DC. Switching on may give additional losses...

**Comment by pvvx** [comment]
st17h66_RF (Lenze_rf.hex) programmed into the AiThinker PB-03F (PHY6252) module
Average consumption ~8.5 uA at 3.3V (120 sec period). Sleep 3.0 uA
![image](https://user-images.githubusercontent.com/12629515/215348908-1cc80781-03e3-4306-8d4e-a09a8d2a0dbd.png)

**Comment by biemster** [comment]
That's looking very good @pvvx , funny that the phy6252 can run on exactly the same firmware! That gives us another platform to test on, while many of us are waiting for the mail to arrive..
I'm trying to remove SRAM retention completely, and have the chip do the whole init every time it wakes up, here:
https://github.com/biemster/st17h66_blinky
this should bring the sleep consumption down another uA, but I think that is really the max I can squeeze out of it.

**Comment by pvvx** [comment]
st17h66_RF (Lenze_rf.hex) programmed into the AiThinker PB-03F (PHY6252) module:
![image](https://user-images.githubusercontent.com/12629515/215352360-7e976eaa-ab77-4aed-8791-acb5859e3ac9.png)
PowerOn:
![image](https://user-images.githubusercontent.com/12629515/215352719-66131eeb-4b39-4330-b94d-f5b9ff299bfb.png)

**Comment by biemster** [comment]
So startup takes 50ms @ 6mA? That would eat the power saving from removing SRAM retention, plus a whole lot more! Do you agree that that is not a viable path, or am I missing something?

**Comment by pvvx** [comment]
[Power Consumption Xiaomi XMWSDJ04MMC](https://pvvx.github.io/XMWSDJ04MMC/power_original.html)
[Power Consumption Xiaomi Mijia BLE MHO-C401 (E-ink)](https://pvvx.github.io/MHO_C401/power_original.html)
[Average power consumption of Original CGG1 ver 2018.11 (Soc: nRF52810).](https://pvvx.github.io/CGG1_old/OriginalPower.html)
[Average power consumption of Original CGG1 "Qingping Temp & RH Monitor" sw ver: 1.0.1_0093.](https://pvvx.github.io/CGG1/OriginalPower.html)
... :)

> So startup takes 50ms @ 6mA?

The measurements are not accurate (+-5% for mA), as they are not calibrated and it is difficult to connect to the Dev-Kit PB-03F.
![image](https://user-images.githubusercontent.com/12629515/215354463-b43eecb8-71a6-481b-a500-5724d0f9df23.png)

And the power to the Dev-Kit from USB makes a lot of noise ...
Incomprehensible impulse:
![image](https://user-images.githubusercontent.com/12629515/215355265-4dc719b2-eeca-4bcd-9af5-c4f854a975aa.png)

**Comment by pvvx** [comment]
> That would eat the power saving from removing SRAM retention, plus a whole lot more!

It is possible that with 'deep-sleep' without 'retention RAM' there will be a long starting period. But it is necessary to check what will be more - addition to 'sleep' or addition to the start of the active period.

ST17H66 is not very suitable for small consumption - 7..9 uA is already good(!).
For example, with EFR32BG22C112  ([Average consumption: 3.2 uA (window: 12 sec, power: 3.3V), advertise interval 1 sec](https://pvvx.github.io/XMWSDJ04MMC/img/power_work.gif)):
**Radio Performance**
• 2.5 mA radio receive current
• 3.4 mA radio transmit current @ 0 dBm output power
• **Low System Energy Consumption**
• 3.6 mA RX current (1 Mbps GFSK)
• 4.1 mA TX current @ 0 dBm output power
• 26 μA/MHz in Active Mode (EM0) at 38.4 MHz
• 1.40 μA EM2 DeepSleep current (32 kB RAM retention and RTC running from LFXO)
• 1.75 μA EM2 DeepSleep current (32 kB RAM retention and RTC running from Precision LFRCO)
• 0.17 μA EM4 current

**Comment by biemster** [comment]
> ST17H66 is not very suitable for small consumption - 7..9 uA is already good(!).

Yeah thanks, I think so too. I'm going to move my attention a bit more to the GCC port. And do a real life test to see how long the CR2023 lasts :)

**Comment by vadimkozhin** [comment]
Hey @biemster Prepare yourself... :) Mine still works from August, without any power optimization.

**Comment by biemster** [comment]
Nice! that would put it on at least half a year and likely a couple months longer, since the sleep draw is now quite a bit lower :+1:

**Comment by sebi5361** [comment]
> > ST17H66 is not very suitable for small consumption - 7..9 uA is already good(!).

Out of curiosity, what would be a BLE SoC suitable for small consumption?

> And do a real life test to see how long the CR2023 lasts :)

...with the RF only firmware or the complete one?

**Comment by biemster** [comment]
> ...with the RF only firmware or the complete one?

The complete one, that one is at least doing something useful :). The RF one does not have the FindMy advertisement implemented yet, since I'm focusing on GCC at the moment.

**Comment by pvvx** [comment]
> And do a real life test to see how long the CR2023 lasts :)

I test all BLE thermometers on CR2016. Some have been with them for over six months. The average consumption with changed settings is 9..14 µA.

**Comment by biemster** [comment]
biemster/st17h66_blinky#2 is sleeping at 0.9uA, but I can't seem to wake it up anymore. For some reason no IRQ handlers are called. @pvvx @sebi5361 @Itheras does any of you maybe have a setup with a debugger attached, to run a quick check on where it blocks? (I unfortunately don't have the hardware :cry:) (On the plus side, blinky compiles with GCC! also with Keil, if you prefer)

**Comment by sebi5361** [comment]
I didn't receive my beacon to help yet ):

I have a few SWD probes (CMSIS-DAP, Black Magic, ST-Link, J-Link). I am used to running OpenOCD.
Are you talking of such a way?
(I don't know if there exists a target in OpenOCD corresponding to the Lenze SoC though.)

**Comment by biemster** [comment]
> I have a few SWD probes (CMSIS-DAP, Black Magic, ST-Link, J-Link). I am used to running OpenOCD.
> Are you talking of such a way?

Yes that! I would just like to see what the cortex-m0 core is doing while sleeping, or waiting for an interrupt. I hope that is standard enough for openocd?

**Comment by sebi5361** [comment]
This I don't know how to do :(
I can unbrick a MCU by wiping it off and reprogramming it through SWD only.
Did you find a way to wake up your MCU?

**Comment by biemster** [comment]
> Did you find a way to wake up your MCU?

As far as I know this requires either an interrupt or event (__WFI() or __WFE()), but blinky doesn't seem to generate, or respond to either at the moment..

**Comment by pvvx** [comment]
Finally came in the mail "iSearching"...
[pvvx.github.io/iSearching](https://pvvx.github.io/iSearching/)

**Comment by pvvx** [comment]
> does any of you maybe have a setup with a debugger attached, to run a quick check on where it blocks?

How will Jtag work during the Sleep SoC if the chip is powered down?
Jtag (SWD) are not suitable for debugging BLE chips.

**Comment by biemster** [comment]
> How will Jtag work during the Sleep SoC if the chip is powered down?

:smile:  that was just to debug the interrupts, they where not working during normal operation either. But I fixed that now biemster/st17h66_blinky#2!

The only issue left for blinky now, is that when the soc goes to sleep it seems to turn off the RTC as well, or the interrupts, or both. Or it is actually waking up, but in a uninitialized state.
![image](https://user-images.githubusercontent.com/5699190/216834237-172040cd-7ec8-443b-9bc8-faa7e4518255.png)
(considering the 1uA draw I measure, the RTC should still be running?)
When that is fixed, I'm going to instead of blink the light, send an advertisement on the phy.

**Comment by pvvx** [comment]
Is this the correct board pinout?
![image](https://user-images.githubusercontent.com/12629515/218384907-f7a33190-b25a-4cc0-b9ef-d276d4fbe859.png)
![image](https://user-images.githubusercontent.com/12629515/218385151-c2885666-e288-4b52-9650-3673bd38f860.png)
https://pvvx.github.io/iSearching/

**Comment by biemster** [comment]
> Is this the correct board pinout?

That looks about right yes. (I only every use p9 and p10, and p3 (and batt))

**Comment by pvvx** [comment]
And how are things with OTA on native firmware?
PhyOTA v2.2.1, PhyApp 1.2.1 is not working.

**Comment by biemster** [comment]
I've seen some references of OTA in the SDK, but I never looked into that at all. Would be quite handy though, although that would require the whole BLE stack and SDK right?

**Comment by pvvx** [comment]
Yes, the entire BLE stack will be needed.
But different bootloader combinations are also possible.
On this key fob, I want to make a switch in HA and a label with the RSA code to open the gate, and so on...

**Comment by biemster** [comment]
That's a very interesting idea! I'm also planning to use them as a simple wireless button, I'll open a repo for this soon. Didn't think much about security though, which would be necessary.

**Comment by biemster** [comment]
@natschil seems to have gotten Zigbee working: https://github.com/biemster/st17h66_RF/issues/7, so I will likely adopt that for my HA switches and ditch BLE in further work.

**Comment by natschil** [comment]
I don't want to hijack the discussion, but I'm having some issues flashing after de-soldering the buzzer - has anyone had the same problem? Previously I was able to flash with UART connected to P9 and P10 on the back of the board..

Edit: it seems that without the buzzer, the board can be powered over the UART pin, so to really reset it it is necessary to disconnect both uart and power.

**Comment by vadimkozhin** [comment]
Some updates, for those who interested. Tag still works. Can see it from the app. Deployed Aug 16, 2022.

**Comment by biemster** [comment]
> Edit: it seems that without the buzzer, the board can be powered over the UART pin, so to really reset it it is necessary to disconnect both uart and power.

yes thanks for reporting this, I'm having issues regularly with this as well when running the flasher

**Comment by biemster** [comment]
> Some updates, for those who interested. Tag still works. Can see it from the app. Deployed Aug 16, 2022.

That's amazing, it's on a 5s advertisement interval right? I really should add battery level reporting to the status byte.

**Comment by vadimkozhin** [comment]
For those who interested: my tag stop transmit on 26.04. So, the tag works from Aug 16, 2022 till Apr 26, 2023. Traveled across EU and stable transmits 24/7, no any freezes or bugs detected during about 8 months.

**Comment by biemster** [comment]
Aww poor thing! But that's a great achievement, I hope some of mine will reach that too.
I very much need to wrap up this power minimization effort, and choose between the SDK code and the st17h66_RF code, especially since quite a few people put significant effort in it.
I hope I'll get some time off work soon for this, and thanks for the update!

**Comment by biemster** [comment]
@vadimkozhin I've been pondering sending a couple out in some sort of art / social experiment, just to see where they travel, has your tag been in a similar project since it traveled a lot?

**Comment by vadimkozhin** [comment]
@biemster I've pass the tag to my friends who has leave for business or travel during last months. Cyprus, Italy, Luthuania, Poland, Netherlands and Germany so far. Works and transmits each time perfectly. I've contact them during travels, and compare the locations. Very accurate results, up until the side ot the street. Amazing!

One thing that suprises and scares a bit — carriers never recieves a single standart Apple warning about 'tag is following you'. I'm having also an original airtag atttached to my dog, but registered not on my id, and it alarms each time when I walk with the pet...

**Comment by biemster** [comment]
That's amazing indeed! And scary! I don't have an iDevice, so I can't corroborate, but I definitely thought that this single key implementation would light up every iPhone that was carried along..

**Comment by Itheras** [comment]
The alerts work for me but with a diminished sensitivity which I don't consider bad . If I am around the block walking I get no alerts but if I am driving for some distance after a certain amount of time I get an alert.

**Comment by vadimkozhin** [comment]
@Itheras Yes, for me is the same. Seems like if the tag which is not registered on same id as the phone travels some distance together the alert appears.

But if the registerd phone travels with this setup, not registerеd phone does not recieve an alert, which is completely fine. And first case is also ok. It prevents attaching the tag to person and spy on them without any notice.

**Comment by olivluca** [comment]
I didn't manage to reprogram the tag yet, but my use case would be to attach it to my car and bicycle, hence I wouldn't want the potential thief to be alerted if he has an iphone.

**Comment by vadimkozhin** [comment]
@olivluca I agree. I just mention it, but is it a bug or a feature depends from the use case 😁

**Comment by Itheras** [comment]
I agree is my experience that the decreased sensitivity make it more suitable for that use. You still get alerts but they behave differently, if attached to a vehicule it takes a longer round trip in a short period for it to trigger in a bicycle it probably won't trigger since it seems the distance is a mayor factor in it triggering. a few miles at least 15+. also even if detected the fact that there is no buzzer to try and locate may spook the thief in to just abandoning what is stolen . in my experience it still has anti stalking functionality but the changed behavior give you more time to track the stolen asset before it gets detected.

**Comment by Blinko1987** [comment]
It’s possible to reduce the size with a dremel tool smaller than the CR2032. Then attach power wires to the VDD cap and use conductive glue to the battery since solder won’t stick. Then harden it all up with epoxy. 
https://postimg.cc/gXHNzQz2

**Comment by johnbaker26222** [comment]
![Uploading A171F5A3-29B2-4E09-96B1-336475BFF202.jpeg…]()
